crates/core/tedge_agent/src/entity_manager/server.rs

```diff
@@ -0,0 +1,187 @@
+use tedge_actors::MessageSink;
+use tedge_actors::Sender;
+use tedge_actors::Server;
+use tedge_api::entity::EntityMetadata;
+use tedge_api::entity_store;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::pending_entity_store::PendingEntityData;
+use tedge_api::EntityStore;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::QoS;
+use tedge_mqtt_ext::TopicFilter;
+use tracing::error;
+
+#[derive(Debug)]
+pub enum EntityStoreRequest {
+    Get(EntityTopicId),
+    Create(EntityRegistrationMessage),
+    Delete(EntityTopicId),
+    MqttMessage(MqttMessage),
+}
+
+#[derive(Debug)]
+pub enum EntityStoreResponse {
+    Get(Option<EntityMetadata>),
+    Create(Result<(Vec<EntityTopicId>, Vec<PendingEntityData>), entity_store::Error>),
+    Delete(Vec<EntityTopicId>),
+    Ok,
+}
+
+pub struct EntityStoreServer {
+    entity_store: EntityStore,
+    mqtt_schema: MqttSchema,
+    mqtt_publisher: LoggingSender<MqttMessage>,
+    entity_auto_register: bool,
+}
+
+impl EntityStoreServer {
+    pub fn new(
+        entity_store: EntityStore,
+        mqtt_schema: MqttSchema,
+        mqtt_actor: &mut impl MessageSink<MqttMessage>,
+        entity_auto_register: bool,
+    ) -> Self {
+        let mqtt_publisher = LoggingSender::new("MqttPublisher".into(), mqtt_actor.get_sender());
+
+        Self {
+            entity_store,
+            mqtt_schema,
+            mqtt_publisher,
+            entity_auto_register,
+        }
+    }
+}
+
+#[async_trait]
+impl Server for EntityStoreServer {
+    type Request = EntityStoreRequest;
+    type Response = EntityStoreResponse;
+
+    fn name(&self) -> &str {
+        "EntityStoreServer"
+    }
+
+    async fn handle(&mut self, request: EntityStoreRequest) -> EntityStoreResponse {
+        match request {
+            EntityStoreRequest::Get(topic_id) => {
+                let entity = self.entity_store.get(&topic_id);
+                EntityStoreResponse::Get(entity.cloned())
+            }
+            EntityStoreRequest::Create(entity) => {
+                let res = self.register_entity(entity).await;
+                EntityStoreResponse::Create(res)
+            }
+            EntityStoreRequest::Delete(topic_id) => {
+                let deleted_entities = self.deregister_entity(topic_id).await;
+                EntityStoreResponse::Delete(deleted_entities)
+            }
+            EntityStoreRequest::MqttMessage(mqtt_message) => {
+                self.process_mqtt_message(mqtt_message).await;
+                EntityStoreResponse::Ok
+            }
+        }
+    }
+}
+
+impl EntityStoreServer {
+    async fn process_mqtt_message(&mut self, message: MqttMessage) {
+        let (topic_id, channel) = self
+            .mqtt_schema
+            .entity_channel_of(&message.topic)
+            .expect("Topic schema must have been pre-validated");
```

Comment:
Actually not safe. Got a panic starting the updated agent in my dev environment (which is not clean, I admit - but this should not crash the agent).

```
thread 'tokio-runtime-worker' panicked at crates/core/tedge_agent/src/entity_manager/server.rs:97:14:
Topic schema must have been pre-validated: Channel(InvalidCategory("cmd/builtin/log_upload/sub:log_upload:c8y-mapper-52201"))
```

Updated; any weird topic makes the agent to crash. => This must be fixed.

Reply:
Resolved by f752408